tools/distpartitioning/multi_dev_init.py

```diff
@@ -0,0 +1,513 @@
+    """
+
+    #determine Global_type_nid for the residing features 
+    start = timer()
+    node_features_rank_lst = []
+    global_nid_rank_lst = []
+    for part_id in np.arange(world_size):
+
+        #form outgoing features to each process
+        send_node_features = {}
+        send_global_nids = {}
+        for ntype_name, ntype_id in ntypes_map.items(): 
+
+            #check if features exist for this node_type
+            if (ntype_name+'/feat' in node_features) and (node_features[ntype_name+'/feat'].shape[0] > 0):
+                feature_count = node_features[ntype_name+'/feat'].shape[0]
+                global_feature_count = ntype_id_count[str(ntype_id)]
+
+                #determine the starting global_nid for this node_type_id
+                feat_per_proc = math.ceil(global_feature_count / world_size)
+                global_type_nid_start = feat_per_proc * rank
+                global_type_nid_end = global_type_nid_start
+                if((global_type_nid_start + feat_per_proc) > global_feature_count):
+                    global_type_nid_end += (ntype_id_count[str(ntype_id)] - global_type_nid_start)
+                    type_nid = np.arange(0, (ntype_id_count[str(ntype_id)] - global_type_nid_start))
+                else: 
+                    global_type_nid_end += feat_per_proc 
+                    type_nid = np.arange(0, feat_per_proc)
+
+                #now map the global_ntype_id to global_nid 
+                global_nid_offset = ntypes_nid_map[ntype_name][0]
+                global_nid_start = global_type_nid_start + global_nid_offset
+                global_nid_end = global_type_nid_end + global_nid_offset
+
+                #assert (global_nid_end - global_nid_start) == feature_count
+                global_nids = np.arange(global_nid_start, global_nid_end, dtype=np.int64)
+
+                #determine node feature ownership 
+                #TODO: a Bug here. 
+                '''
+                part_ids = node_part_ids[global_nids] 
+                idx = (part_ids == part_id)
+                out_global_nid = global_nids[idx == 1]
+                out_type_nid = type_nid[idx == 1]
+                out_features = node_features[ntype_name+'/feat'][out_type_nid]
+                send_node_features[ntype_name+'/feat'] = out_features
+                send_global_nids[ntype_name+'/feat'] = out_global_nid
+                '''
+                part_ids_slice = node_part_ids[global_nid_start:global_nid_end]
+                idx = (part_ids_slice == part_id)
+                out_global_nid = global_nids[idx == 1]
+                out_type_nid = type_nid[idx == 1]
+                out_features = node_features[ntype_name+'/feat'][out_type_nid]
+                send_node_features[ntype_name+'/feat'] = out_features
+                send_global_nids[ntype_name+'/feat'] = out_global_nid
+
+        node_features_rank_lst.append(send_node_features)
```

Comment:
It seems we need to cache all the `send_node_features`s from different node types. Will this consume lot of memory?